File: 1_fibonacci_number/Fibonacci.java
/*
 * Author: Jehadur Rahman
 * Organization: CyArm
 * GitHub: https://github.com/JehadurRE
 * Website: jehadurre.me
 * Date: 2025-07-10 09:43 (Thu)
 */

import java.util.Scanner;

public class Fibonacci {
  private static long calc_fib_naive(int n) {
    if (n <= 1)
      return n;

    return calc_fib_naive(n - 1) + calc_fib_naive(n - 2);
  }

  private static long calc_fib_fast(int n) {
    if (n <= 1)
      return n;

    long[] arr = new long[n + 1];
    arr[0] = 0;
    arr[1] = 1;

    for (int i = 2; i <= n; i++) {
      arr[i] = arr[i - 1] + arr[i - 2];
    }

    return arr[n];
  }

  private static void test_solution() {
    assert calc_fib_fast(3) == 2;
    assert calc_fib_fast(10) == 55;
    for (int n = 0; n < 20; n++) {
      assert calc_fib_fast(n) == calc_fib_naive(n);
    }
  }

  public static void main(String args[]) {
    Scanner in = new Scanner(System.in);
    int n = in.nextInt();

    // System.out.println(calc_fib_naive(n));  // Slow version
    // test_solution();  // Test the implementation
    System.out.println(calc_fib_fast(n));
  }
}

File: 1_fibonacci_number/Fibonacci.kt
/*
 * Author: Jehadur Rahman
 * Organization: CyArm
 * GitHub: https://github.com/JehadurRE
 * Website: jehadurre.me
 * Date: 2025-07-10 09:43 (Thu)
 */

import java.util.*

fun calcFibNaive(n: Long): Long {
    return if (n <= 1) n else calcFibNaive(n - 1) + calcFibNaive(n - 2)
}

fun calcFib(n: Long): Long {
    if (n <= 1) return n
    
    val arr = LongArray(n.toInt() + 1)
    arr[0] = 0
    arr[1] = 1
    
    for (i in 2..n.toInt()) {
        arr[i] = arr[i - 1] + arr[i - 2]
    }
    
    return arr[n.toInt()]
}

fun testSolution() {
    assert(calcFib(3) == 2L)
    assert(calcFib(10) == 55L)
    for (n in 0L..19L) {
        assert(calcFib(n) == calcFibNaive(n))
    }
}

fun main(args: Array<String>) {
    val scanner = Scanner(System.`in`)
    val n = scanner.nextLong()

    // println(calcFibNaive(n))  // Slow version
    // testSolution()  // Test the implementation
    println(calcFib(n))
}